README.md
# Python client for Strava Flyby

[![PyPI version](https://badge.fury.io/py/strava-flyby.svg)](https://badge.fury.io/py/strava-flyby)
[![Build Status](https://travis-ci.org/sladkovm/strava-flyby.svg?branch=master)](https://travis-ci.org/sladkovm/strava-flyby)


## Install

$ pip install strava-flyby


## Usage

*strava-flyby* provides a convenience object wrapping of the *Strava Labs Flyby API* response. Next to providing a direct access to the unmodified content of the *Flyby API* response, the object will expose number of convenience attributes and methods to access the list of *activity*, *ids*, *matches* and *athletes*.  


Jupyter notebook example with real data is [here](https://github.com/sladkovm/strava-flyby/blob/master/examples/Flyby%20quickstart.ipynb)


Create a *flyby* object using the Strava activity ID with a help of the `flyby()` factory function:
```python
from flyby import flyby
fb = flyby(activity_id=12345)
fb
>>> Flyby object with 17 matches. 
        Attributes: ids, activity, matches, athletes
        Methods: matches_to_list(), matches_to_json(), get_ids()
```

Filter by distance in km and print all matched ids:
```python
print(fb.get_ids(distance=(110, 130)))
```

Dump filtered by distance flattened matches into a list:
```python
print(fb.matches_to_list(distance=120, tol=0.1))
```


CHANGES.rst
Changelog
=========

0.0.6 (2019-07-01)
------------------

Breaking changes:

- Attributes raw_activity, raw_matches, raw_athletes are accessed as activity, matches and athletes respectively

New features:

- added __str__ and __repr__ to Flyby object

Bug fixes:

- Fixed requests to Strava Labs



0.0.5 (2018-12-28)
------------------

Bug fixes:

- updated dependencies

0.0.4 (2018-12-28)
------------------

Bug fixes:

- updated dependencies

0.0.3 (2018-09-20)
------------------

Breaking changes:

- ids becomes and attribute wrapper
- get_ids() is a full function to filter the ids
- explicit prefixes raw_ for all attributes that return original data or their
subset
- matches is an attribute wrapper that returns the flattened ready to use in
DataFrame list of matched activities

New features:

- pipenv
- travisio

Bug fixes:

- *add item here*


0.0.2 (2018-03-05)
------------------

Breaking changes:

- *add item here*

New features:

- to_list and to_json methods to dump flattened matches
- distance filter by center distance with +/- tolerance
- distance filter by range (start, end)

Bug fixes:

- distance filter


0.0.1 (2018-03-04)
------------------

Breaking changes:

- *add item here*

New features:

- First release

Bug fixes:

- *add item here*



0.X.X (unreleased)
------------------

Breaking changes:

- *add item here*

New features:

- Stress in zone

Bug fixes:

- *add item here*


tests/test_flyby.py
import pytest
import json
import os
import httpretty
from flyby import Flyby, flyby, _request_strava_flyby


@pytest.fixture
def flyby_response():

    current_dir = os.path.dirname(os.path.abspath(__file__))
    f_name = os.path.join(current_dir, 'flyby.json')

    with open(f_name, encoding='utf-8') as f:

        rv = json.dumps(json.load(f))

    return rv


def test_request_strava_flyby():
    id = 2488860384
    content = _request_strava_flyby(id)
    assert content is not None
    assert 'activity' in content.keys()
    assert 'matches' in content.keys()
    assert 'athletes' in content.keys()
    assert len(content['matches']) > 1


def test_constructor_with_none():

    fb = Flyby()

    assert fb.raw_content is None
    assert fb.activity is None
    assert fb.matches is None
    assert fb.athletes is None
    assert fb.ids is None
    assert fb.matches is None


def test_constructor_with_real_data(flyby_response):

    expected = json.loads(flyby_response)

    fb = Flyby(expected)

    assert fb.raw_content == expected
    assert fb.activity == expected['activity']
    assert fb.matches == expected['matches']
    assert fb.athletes == expected['athletes']
    assert len(fb.ids) == 751
    assert len(fb.matches) == 751

    assert fb._distance_filter().sum() == 751
    assert fb._distance_filter(distance=100).sum() == 222
    assert fb._distance_filter(distance=100, tol=0.05).sum() == 68
    assert fb._distance_filter(distance=(90, 110)).sum() == 222

    assert len(fb.get_ids()) == 751
    assert len(fb.get_ids(distance=100)) == 222
    assert len(fb.get_ids(distance=100, tol=0.05)) == 68
    assert len(fb.get_ids(distance=(90, 110), tol=0.05)) == 222

    assert len(fb.matches_to_json()) == 751
    assert len(fb.matches_to_json(distance=100)) == 222

    assert len(fb.matches_to_list()) == 751
    assert len(fb.matches_to_list(distance=100)) == 222


@httpretty.activate
def test_flyby(flyby_response):

    test_id = 12345
    httpretty.register_uri(httpretty.GET,
                           "https://nene.strava.com/flyby/matches/{}".format(test_id),
                           body=flyby_response,
                           content_type="application/json")

    expected = json.loads(flyby_response)

    fb = flyby(test_id)

    assert fb.raw_content == expected
    assert fb.activity == expected['activity']
    assert fb.matches == expected['matches']
    assert fb.athletes == expected['athletes']


flyby.py
"""
1. Get IDs from flyby:
    - distance
    - elapsedTime
    - startTime

>>> fb = flyby(activity_id=12345)

"""
import os
import requests
import json
import pandas as pd
from loguru import logger


def flyby(activity_id):
    """Find flybys for a given activity

    Parameters
    ----------
    activity_id : int
        Strava activity id

    Returns
    -------
    obj : Flyby
    """
    return Flyby(_request_strava_flyby(activity_id))


def _request_strava_flyby(activity_id):
    """Request the Strava Labs Flyby and return the json content
    
    Returns:
    --------
    dict with keys 'activity', 'matches', 'athletes'
    """
    __base_url = 'https://nene.strava.com/flyby/matches/'
    __headers = {'referer': 'https://labs.strava.com/flyby/viewer/',
                 'origin': 'https://labs.strava.com'}
    logger.info("Requesting Strava Flyby for activity ID = {}".format(activity_id))
    r = requests.get('{}{}'.format(__base_url, activity_id), headers=__headers)
    if r.ok:
        content = json.loads(r.text)
        return content
    else:
        raise ConnectionError("Flyby returned {}".format(r.status_code))


class Flyby():
    """Results of Flyby search


    Attributes
    ----------
    raw_content: dict
        Deserialized and unprocessed JSON reponse of the Strava Flyby API

    """

    def __init__(self, content=None):
        """Create Flyby object

        This objects serves as a convenience wrapper
        around the Strava Flyby API JSON response. The easiest way to initialite is
        to call the factory function flyby(activity_id).

        Parameters
        ----------
        raw_content : dict
            Deserialized and unprocessed JSON response of the Strava Flyby API
        raw_activity: dict
            Requesting activity
        raw_matches: list
            Found list of matches
        matches: List
            Flattened list of matches ready to use for DataFrame
        """

        self.raw_content = content


    def __str__(self):
        return f"{self.raw_content}"


    def __repr__(self):
        n_matches = len(self.matches)
        return f"""Flyby object with {n_matches} matches. 
        Attributes: ids, activity, matches, athletes
        Methods: matches_to_list(), matches_to_json()"""


    def matches_to_list(self, **kwargs):
        """Dump flattened matches into a list

        Parameters
        ----------
        kwargs : see to_json

        Returns
        -------
        list
        """

        return self.matches_to_json(**kwargs)


    def matches_to_json(self, path_or_buf=None, **kwargs):
        """Dump flattened matches into json or a list

        Parameters
        ----------
        path_or_buf : file path, optional
            default=None, which means the results will be dumped into a list of dicts
        distance : number, optional
            Distance in km, default=None, which means all ids are returned
        tol : float from 0 to 1, optional
            Tolerance +/- on `distance`, only applicable if former is not None, default=0.1

        Returns
        -------
        None or list
        """

        df = pd.DataFrame(self._flatten_matches())

        rv = df[self._distance_filter(**kwargs)]

        if path_or_buf:

            rv.to_json(path_or_buf, orient=kwargs.get('orient', 'records'))

        else:

            return json.loads(rv.to_json(orient=kwargs.get('orient', 'records')))


    def get_ids(self, **kwargs):
        """IDs returned by Flyby

        Parameters
        ----------
        distance : number, optional
            Distance in km, default=None, which means all ids are returned
        tol : float from 0 to 1, optional
            Tolerance +/- on `distance`, only applicable if former is not None, default=0.1

        Returns
        -------
        list
        """

        df = pd.DataFrame(self._flatten_matches())

        return df[self._distance_filter(**kwargs)]['id'].tolist()



    def _distance_filter(self, **kwargs):
        """

        Parameters
        ----------
        distance : number or a tuple, optional
            If number, than distance is treated as a center distance and the range
            is calculated as `distance` +/- `tol`*`distance`
            If tuple, than the tuple is treated as a range
        tol : float, optional
            Only relevant with center distance, default=0.1

        Returns
        -------
        pd.Series
        """

        df = pd.DataFrame(self._flatten_matches())

        if not kwargs.get('distance', None):

            return df['distance'] > -1

        if (type(kwargs.get('distance')) == int) or (type(kwargs.get('distance')) == float):

            rv = ((df['distance'] > (1.0 - kwargs.get('tol', 0.1)) * kwargs.get('distance') * 1000) &
                  (df['distance'] < (1.0 + kwargs.get('tol', 0.1)) * kwargs.get('distance') * 1000))

        elif type(kwargs.get('distance')) == tuple:

            rv = ((df['distance'] > kwargs.get('distance')[0] * 1000) &
                  (df['distance'] < kwargs.get('distance')[1] * 1000))

        else:

            raise ValueError("Distance must be either a number or a tuple of numbers")

        return rv


    def _flatten_matches(self):
        """Convert all matches into flat dict

        Returns
        -------
        dict
            With keys: ['activityType', 'athleteId', 'closestDistance', 'closestPoint',
                'correlation', 'distance', 'elapsedTime', 'id', 'name', 'spatialCorrelation', 'startTime' ]
        """

        rv = []
        for m in self.matches:

            _a = m['otherActivity']
            _c = m['correlation']
            _a.update(_c)

            rv.append(_a)

        return rv

    @property
    def ids(self):
        """List of found activities ids

        Returns
        -------
        list
        """
        if self.raw_content:
            return self.get_ids()
        else:
            return None

    @property
    def activity(self):
        """Requesting activity

        Returns
        -------
        dict
        """
        if self.raw_content:
            return self.raw_content.get('activity', None)
        else:
            return None

    @property
    def matches(self):
        """Flattend matches

        Returns
        -------
        list of dict ready to use as input for DataFrame
        """

        if self.raw_content:
            return self.matches_to_list()
        else:
            return None

    @property
    def matches(self):
        """Matches returned by Flyby

        Returns
        -------
        list
        """
        if self.raw_content:
            return self.raw_content.get('matches', None)
        else:
            return None


    @property
    def athletes(self):
        """Athletes returned by Flyby

        Returns
        -------
        dict
        """
        if self.raw_content:
            return self.raw_content.get('athletes', None)
        else:
            return None
